File: backend/app/Models/Contract.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Carbon\Carbon;
use Illuminate\Database\Eloquent\Relations\HasOne;

class Contract extends Model
{
    use HasFactory;

    protected $fillable = [
        'employee_id',
        'type',
        'salary_id',
        'contract_code',
        'start_date',
        'end_date',
        'sign_date',
        'contract_file',
        'status',
        'note',
        'year_of_service',
    ];

    public function employee(): BelongsTo
    {
        return $this->belongsTo(Employee::class, 'employee_id');
    }

    public function contractType(): BelongsTo
    {
        return $this->belongsTo(ContractType::class, 'type');
    }

    public function salary(): BelongsTo
    {
        return $this->belongsTo(Salary::class, 'salary_id');
    }

    public function positions(): BelongsToMany
    {
        return $this->belongsToMany(Position::class, 'contract_employee_position', 'contract_id', 'position_id')
            ->withPivot(['is_main', 'start_date', 'end_date', 'base_salary', 'ratio']);
    }

    public function getYearOfServiceAttribute()
    {
        $start = Carbon::parse($this->start_date);
        // Nếu end_date không tồn tại (null), nghĩa là hợp đồng vẫn còn hiệu lực
        if ($this->end_date) {
            $end = Carbon::parse($this->end_date);
            return $start->diffInYears($end);
        } else {
            return $start->diffInYears(Carbon::now()); // Tính đến năm hiện tại
        }
    }
}

File: hr_management_app/lib/models/contract.dart
class Contract {
  final String contractCode;
  final int employeeId;
  final int salaryId;
  final int type;
  final String startDate;
  final String? endDate;
  final String signDate;
  final String? contractFile;
  final String status;
  final String? note;
  final int yearOfService;
  final String createdAt;
  final String updatedAt;

  Contract({
    required this.contractCode,
    required this.employeeId,
    required this.salaryId,
    required this.type,
    required this.startDate,
    this.endDate,
    required this.signDate,
    this.contractFile,
    required this.status,
    this.note,
    required this.yearOfService,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Contract.fromJson(Map<String, dynamic> json) {
    return Contract(
      contractCode: json['contract_code'] ?? '',
      employeeId: json['employee_id'] ?? 0,
      salaryId: json['salary_id'] ?? 0,
      type: json['type'] ?? 0,
      startDate: json['start_date'] ?? '',
      endDate: json['end_date'],
      signDate: json['sign_date'] ?? '',
      contractFile: json['contract_file'],
      status: json['status'] ?? '',
      note: json['note'],
      yearOfService: json['year_of_service'] ?? 0,
      createdAt: json['created_at'] ?? '',
      updatedAt: json['updated_at'] ?? '',
    );
  }
}